File: pkg/peer/peer.go
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

package peer

import (
	"bytes"
	"encoding/binary"
	"encoding/hex"
	"errors"
	"fmt"
	"os"
	"runtime"
	"sort"
	"time"
	"unicode/utf8"
	"unsafe"
)

const (
	peerSendErrInvalidPeerCount = -1
	peerSendErrDataTooLarge     = -2
	peerSendErrChannelFull      = -3
	peerSendErrChannelClosed    = -4
	peerSendErrInvalidPeerID    = -5

	peerRecvErrChannelDisconnected = -1
)

//go:wasmimport terminal_games peer_send
//go:noescape
func peer_send(peer_ids_ptr unsafe.Pointer, peer_ids_count uint32, data_ptr unsafe.Pointer, data_len uint32) int32

//go:wasmimport terminal_games peer_recv
//go:noescape
func peer_recv(from_peer_ptr unsafe.Pointer, data_ptr unsafe.Pointer, data_max_len uint32) int32

//go:wasmimport terminal_games node_latency
//go:noescape
func node_latency(node_ptr unsafe.Pointer) int32

//go:wasmimport terminal_games peer_list
//go:noescape
func peer_list(peer_ids_ptr unsafe.Pointer, max_length uint32, total_count_ptr unsafe.Pointer) int32

// ID represents a peer identifier
type ID struct {
	node       NodeID
	timestamp  uint64
	randomness uint32
}

type NodeID [4]byte

const (
	nodeIDByteLen = 4
	idByteLen     = 16
)

func (r NodeID) String() string {
	return string(r[:])
}

var (
	ErrLatencyUnknown      = errors.New("latency unknown")
	ErrInvalidPeerCount    = errors.New("invalid peer count (must be 1-1024)")
	ErrDataTooLarge        = errors.New("data too large: maximum 64KB")
	ErrChannelFull         = errors.New("send channel full, message dropped")
	ErrChannelClosed       = errors.New("send channel closed")
	ErrInvalidPeerID       = errors.New("invalid peer ID: peer ID contains invalid node bytes")
	ErrChannelDisconnected = errors.New("receive channel disconnected")
	ErrListFailed          = errors.New("peer_list failed")
)

// Latency returns the current latency to this node in milliseconds.
func (r NodeID) Latency() (uint32, error) {
	ms := node_latency(unsafe.Pointer(&r[0]))
	if ms < 0 {
		return 0, ErrLatencyUnknown
	}
	return uint32(ms), nil
}

// String returns the human-readable node-timestamp-randomness representation of the ID.
func (id ID) String() string {
	return fmt.Sprintf("%s-%d-%d", id.node, id.timestamp, id.randomness)
}

// ParseID parses a hex-encoded string into an ID
func ParseID(s string) (ID, error) {
	var id ID
	decoded, err := hex.DecodeString(s)
	if err != nil {
		return id, fmt.Errorf("failed to decode hex: %w", err)
	}
	if len(decoded) != idByteLen {
		return id, fmt.Errorf("invalid ID length: expected %d bytes, got %d", idByteLen, len(decoded))
	}
	var bytes [idByteLen]byte
	copy(bytes[:], decoded)
	id = idFromBytes(bytes)
	if !utf8.Valid(id.node[:]) {
		return ID{}, ErrInvalidPeerID
	}
	return id, nil
}

// Timestamp returns the time the peer ID was created
func (id ID) Timestamp() time.Time {
	ms := id.timestamp
	return time.Unix(int64(ms/1000), int64(ms%1000)*int64(time.Millisecond))
}

// Randomness returns the randomness component of the peer ID
func (id ID) Randomness() uint32 {
	return id.randomness
}

// Node returns the node component of the peer ID
func (id ID) Node() NodeID {
	return id.node
}

// Latency returns the current latency to this peer in milliseconds
func (id ID) Latency() (uint32, error) {
	return id.Node().Latency()
}

// Send is shorthand for `peer.Send(data, id)`
func (id ID) Send(data []byte) error {
	return Send(data, id)
}

func CurrentID() ID {
	// PEER_ID is always defined as a valid peer id, so we can ignore the error
	id, _ := ParseID(os.Getenv("PEER_ID"))
	return id
}

// List returns all peers currently connected to this app across all nodes.
// This includes the current peer.
func List() ([]ID, error) {
	for {
		preCount, err := Count()
		if err != nil {
			return nil, err
		}
		peers, totalCount, err := listN(preCount)
		if err != nil {
			return nil, err
		}
		if preCount == totalCount {
			return peers, nil
		}
		runtime.Gosched()
	}
}

func listN(length uint32) ([]ID, uint32, error) {
	var totalCount uint32

	if length == 0 {
		ret := peer_list(nil, 0, unsafe.Pointer(&totalCount))
		if ret < 0 {
			return nil, 0, ErrListFailed
		}
		return nil, totalCount, nil
	}

	buf := make([]byte, length*idByteLen)
	ret := peer_list(unsafe.Pointer(&buf[0]), length, unsafe.Pointer(&totalCount))
	if ret < 0 {
		return nil, totalCount, ErrListFailed
	}

	count := int(ret)
	peers := make([]ID, count)
	for i := range peers {
		offset := i * idByteLen
		var peerBytes [idByteLen]byte
		copy(peerBytes[:], buf[offset:offset+idByteLen])
		peers[i] = idFromBytes(peerBytes)
	}

	return peers, totalCount, nil
}

// Count returns the total number of peers connected to this app without fetching the list.
func Count() (uint32, error) {
	_, totalCount, err := listN(0)
	if err != nil {
		return 0, err
	}
	return totalCount, nil
}

// Message represents a message received from a peer
type Message struct {
	From ID
	Data []byte
}

// Send sends data to one or more peers
func Send(data []byte, peerIDs ...ID) error {
	return SendTo(data, peerIDs)
}

// SendTo sends data to one or more peers
func SendTo(data []byte, peerIDs []ID) error {
	if len(peerIDs) == 0 || len(peerIDs) > 1024 {
		return ErrInvalidPeerCount
	}
	if len(data) > 64*1024 {
		return ErrDataTooLarge
	}

	peerIDsBuf := make([]byte, len(peerIDs)*idByteLen)
	for i, id := range peerIDs {
		peerBytes := id.toBytes()
		copy(peerIDsBuf[i*idByteLen:(i+1)*idByteLen], peerBytes[:])
	}

	var dataPtr unsafe.Pointer
	if len(data) > 0 {
		dataPtr = unsafe.Pointer(&data[0])
	}

	ret := peer_send(
		unsafe.Pointer(&peerIDsBuf[0]),
		uint32(len(peerIDs)),
		dataPtr,
		uint32(len(data)),
	)
	if ret < 0 {
		return sendErrorFromCode(ret)
	}

	return nil
}

func sendErrorFromCode(code int32) error {
	switch code {
	case peerSendErrInvalidPeerCount:
		return ErrInvalidPeerCount
	case peerSendErrDataTooLarge:
		return ErrDataTooLarge
	case peerSendErrChannelFull:
		return ErrChannelFull
	case peerSendErrChannelClosed:
		return ErrChannelClosed
	case peerSendErrInvalidPeerID:
		return ErrInvalidPeerID
	default:
		return fmt.Errorf("unknown peer_send error: %d", code)
	}
}

// Recv waits for a message from any peer
func Recv() (Message, error) {
	fromPeerBuf := make([]byte, idByteLen)
	dataBuf := make([]byte, 64*1024)

	for {
		ret := peer_recv(
			unsafe.Pointer(&fromPeerBuf[0]),
			unsafe.Pointer(&dataBuf[0]),
			uint32(len(dataBuf)),
		)

		if ret > 0 {
			var fromPeerBytes [idByteLen]byte
			copy(fromPeerBytes[:], fromPeerBuf)
			return Message{
				From: idFromBytes(fromPeerBytes),
				Data: dataBuf[:ret],
			}, nil
		} else if ret == 0 {
			time.Sleep(10 * time.Millisecond)
			continue
		} else {
			return Message{}, recvErrorFromCode(ret)
		}
	}
}

func recvErrorFromCode(code int32) error {
	switch code {
	case peerRecvErrChannelDisconnected:
		return ErrChannelDisconnected
	default:
		return fmt.Errorf("unknown peer_recv error: %d", code)
	}
}

type ByPeerId []ID

var _ sort.Interface = ByPeerId(nil)

func (p ByPeerId) Len() int      { return len(p) }
func (p ByPeerId) Swap(i, j int) { p[i], p[j] = p[j], p[i] }
func (p ByPeerId) Less(i, j int) bool {
	leftNode := p[i].node
	rightNode := p[j].node
	if leftNode != rightNode {
		return bytes.Compare(leftNode[:], rightNode[:]) < 0
	}

	leftTimestamp := p[i].timestamp
	rightTimestamp := p[j].timestamp
	if leftTimestamp != rightTimestamp {
		return leftTimestamp < rightTimestamp
	}

	return p[i].randomness < p[j].randomness
}

func idFromBytes(bytes [idByteLen]byte) ID {
	var node NodeID
	copy(node[:], bytes[:nodeIDByteLen])
	return ID{
		node:       node,
		timestamp:  binary.BigEndian.Uint64(bytes[nodeIDByteLen:12]),
		randomness: binary.BigEndian.Uint32(bytes[12:idByteLen]),
	}
}

func (id ID) toBytes() [idByteLen]byte {
	var bytes [idByteLen]byte
	copy(bytes[:nodeIDByteLen], id.node[:])
	binary.BigEndian.PutUint64(bytes[nodeIDByteLen:12], id.timestamp)
	binary.BigEndian.PutUint32(bytes[12:idByteLen], id.randomness)
	return bytes
}

File: terminal-games-sdk/src/peer.rs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

use std::str::FromStr;
use std::time::{SystemTime, UNIX_EPOCH};

const PEER_SEND_ERR_INVALID_PEER_COUNT: i32 = -1;
const PEER_SEND_ERR_DATA_TOO_LARGE: i32 = -2;
const PEER_SEND_ERR_CHANNEL_FULL: i32 = -3;
const PEER_SEND_ERR_CHANNEL_CLOSED: i32 = -4;
const PEER_SEND_ERR_INVALID_PEER_ID: i32 = -5;

const PEER_RECV_ERR_CHANNEL_DISCONNECTED: i32 = -1;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct NodeId([u8; 4]);

const NODE_ID_ERROR: &str = "node id must be valid UTF-8 and occupy exactly 4 bytes";

impl NodeId {
    pub const BYTE_LEN: usize = 4;

    pub fn as_str(&self) -> &str {
        std::str::from_utf8(&self.0).expect("NodeId invariant violated")
    }

    /// Returns the current latency to this node in milliseconds.
    /// Returns `None` if the latency is unknown or the node is not connected.
    pub fn latency(&self) -> Option<u32> {
        let ret = unsafe { crate::internal::node_latency(self.0.as_ptr()) };
        if ret < 0 { None } else { Some(ret as u32) }
    }
}

impl std::fmt::Display for NodeId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.as_str())
    }
}

impl TryFrom<[u8; 4]> for NodeId {
    type Error = PeerError;

    fn try_from(bytes: [u8; 4]) -> Result<Self, Self::Error> {
        std::str::from_utf8(&bytes).map_err(|_| PeerError::InvalidId(NODE_ID_ERROR))?;
        Ok(Self(bytes))
    }
}

impl FromStr for NodeId {
    type Err = PeerError;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        let bytes: [u8; 4] = value
            .as_bytes()
            .try_into()
            .map_err(|_| PeerError::InvalidId(NODE_ID_ERROR))?;
        Self::try_from(bytes)
    }
}

impl From<NodeId> for [u8; 4] {
    fn from(node_id: NodeId) -> Self {
        node_id.0
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct PeerId {
    node: NodeId,
    timestamp: u64,
    randomness: u32,
}

const PEER_ID_ERROR: &str = "peer ID contains invalid node bytes";

impl PeerId {
    pub const BYTE_LEN: usize =
        NodeId::BYTE_LEN + std::mem::size_of::<u64>() + std::mem::size_of::<u32>();

    pub fn to_bytes(&self) -> [u8; Self::BYTE_LEN] {
        let mut bytes = [0u8; Self::BYTE_LEN];
        bytes[0..NodeId::BYTE_LEN].copy_from_slice(&<[u8; NodeId::BYTE_LEN]>::from(self.node));
        bytes[NodeId::BYTE_LEN..12].copy_from_slice(&self.timestamp.to_be_bytes());
        bytes[12..Self::BYTE_LEN].copy_from_slice(&self.randomness.to_be_bytes());
        bytes
    }

    /// Returns the time the peer ID was created
    pub fn timestamp(&self) -> SystemTime {
        let ms = self.timestamp;
        UNIX_EPOCH + std::time::Duration::from_millis(ms)
    }

    /// Returns the randomness component of the peer ID
    pub fn randomness(&self) -> u32 {
        self.randomness
    }

    /// Returns the node component of the peer ID
    pub fn node(&self) -> NodeId {
        self.node
    }

    /// Sends data to this peer
    pub fn send(&self, data: &[u8]) -> Result<(), PeerError> {
        send(data, &[*self])
    }

    /// Returns the current latency to the peer in milliseconds.
    /// Returns `None` if the latency is unknown
    pub fn latency(&self) -> Option<u32> {
        self.node().latency()
    }
}

impl TryFrom<[u8; 16]> for PeerId {
    type Error = PeerError;

    fn try_from(bytes: [u8; 16]) -> Result<Self, Self::Error> {
        let node_bytes: [u8; NodeId::BYTE_LEN] = bytes[0..NodeId::BYTE_LEN].try_into().unwrap();
        let node = NodeId::try_from(node_bytes)
            .map_err(|_| PeerError::InvalidId(PEER_ID_ERROR))?;
        let timestamp = u64::from_be_bytes(bytes[NodeId::BYTE_LEN..12].try_into().unwrap());
        let randomness = u32::from_be_bytes(bytes[12..Self::BYTE_LEN].try_into().unwrap());
        Ok(Self {
            node,
            timestamp,
            randomness,
        })
    }
}

impl std::fmt::Display for PeerId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}-{}-{}", self.node, self.timestamp, self.randomness)
    }
}

impl FromStr for PeerId {
    type Err = PeerError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        parse_id(s)
    }
}

/// Parses a hex-encoded string into a peer ID
pub fn parse_id(s: &str) -> Result<PeerId, PeerError> {
    if s.len() != PeerId::BYTE_LEN * 2 {
        return Err(PeerError::InvalidId("expected 32 hex characters"));
    }
    let mut id = [0u8; PeerId::BYTE_LEN];
    for (i, chunk) in s.as_bytes().chunks(2).enumerate() {
        if chunk.len() != 2 {
            return Err(PeerError::InvalidId("invalid hex string"));
        }
        let byte = u8::from_str_radix(
            std::str::from_utf8(chunk).map_err(|_| PeerError::InvalidId("invalid UTF-8"))?,
            16,
        )
        .map_err(|_| PeerError::InvalidId("failed to decode hex"))?;
        id[i] = byte;
    }
    PeerId::try_from(id)
}

pub fn current_id() -> PeerId {
    // PEER_ID is always defined as a valid peer id, so we can ignore the error
    parse_id(&std::env::var("PEER_ID").unwrap()).unwrap()
}

/// Returns all peers currently connected to this app across all nodes,
/// including the current peer.
///
/// Note: this list is _eventually consistent_ and unordered, meaning that each
/// instance of your app may not see the same list at the same time or in the
/// same order
pub fn list() -> Result<Vec<PeerId>, PeerError> {
    loop {
        let pre_count = count()?;
        let (peers, total_count) = list_n(pre_count)?;
        if pre_count == total_count {
            return Ok(peers);
        }
    }
}

fn list_n(length: u32) -> Result<(Vec<PeerId>, u32), PeerError> {
    let mut total_count: u32 = 0;

    if length == 0 {
        let ret = unsafe { crate::internal::peer_list(std::ptr::null_mut(), 0, &mut total_count) };
        if ret < 0 {
            return Err(PeerError::ListFailed);
        }
        return Ok((Vec::new(), total_count));
    }

    let mut buf = vec![0u8; length as usize * PeerId::BYTE_LEN];

    let ret = unsafe { crate::internal::peer_list(buf.as_mut_ptr(), length, &mut total_count) };

    if ret < 0 {
        return Err(PeerError::ListFailed);
    }

    let peers_vec = buf[..ret as usize * PeerId::BYTE_LEN]
        .chunks_exact(PeerId::BYTE_LEN)
        .map(|chunk| {
            let peer_bytes: [u8; PeerId::BYTE_LEN] = chunk.try_into().unwrap();
            PeerId::try_from(peer_bytes)
        })
        .collect::<Result<Vec<_>, _>>()?;

    Ok((peers_vec, total_count))
}

/// Returns the total count of peers connected to this app without fetching the
/// list.
pub fn count() -> Result<u32, PeerError> {
    let (_, total_count) = list_n(0)?;
    Ok(total_count)
}

/// A message received from a peer
#[derive(Clone, Debug)]
pub struct Message {
    pub from: PeerId,
    pub data: Vec<u8>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PeerError {
    InvalidId(&'static str),
    InvalidPeerCount,
    DataTooLarge,
    ChannelFull,
    ChannelClosed,
    ChannelDisconnected,
    ListFailed,
    Unknown(i32),
}

impl PeerError {
    fn from_send_code(code: i32) -> Self {
        match code {
            PEER_SEND_ERR_INVALID_PEER_COUNT => PeerError::InvalidPeerCount,
            PEER_SEND_ERR_DATA_TOO_LARGE => PeerError::DataTooLarge,
            PEER_SEND_ERR_CHANNEL_FULL => PeerError::ChannelFull,
            PEER_SEND_ERR_CHANNEL_CLOSED => PeerError::ChannelClosed,
            PEER_SEND_ERR_INVALID_PEER_ID => PeerError::InvalidId(PEER_ID_ERROR),
            _ => PeerError::Unknown(code),
        }
    }

    #[allow(unused)]
    fn from_recv_code(code: i32) -> Self {
        match code {
            PEER_RECV_ERR_CHANNEL_DISCONNECTED => PeerError::ChannelDisconnected,
            _ => PeerError::Unknown(code),
        }
    }
}

impl std::fmt::Display for PeerError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            PeerError::InvalidId(msg) => write!(f, "invalid peer ID: {}", msg),
            PeerError::InvalidPeerCount => write!(f, "invalid peer count (must be 1-1024)"),
            PeerError::DataTooLarge => write!(f, "data too large: maximum 64KB"),
            PeerError::ChannelFull => write!(f, "send channel full, message dropped"),
            PeerError::ChannelClosed => write!(f, "send channel closed"),
            PeerError::ChannelDisconnected => write!(f, "receive channel disconnected"),
            PeerError::ListFailed => write!(f, "peer_list failed"),
            PeerError::Unknown(code) => write!(f, "unknown error: {}", code),
        }
    }
}

impl std::error::Error for PeerError {}

/// Sends data to one or more peers
pub fn send(data: &[u8], peer_ids: &[PeerId]) -> Result<(), PeerError> {
    if peer_ids.is_empty() || peer_ids.len() > 1024 {
        return Err(PeerError::InvalidPeerCount);
    }
    if data.len() > 64 * 1024 {
        return Err(PeerError::DataTooLarge);
    }

    let mut peer_ids_buf = Vec::with_capacity(peer_ids.len() * PeerId::BYTE_LEN);
    for id in peer_ids {
        peer_ids_buf.extend_from_slice(&id.to_bytes());
    }

    let data_ptr = if data.is_empty() {
        std::ptr::null()
    } else {
        data.as_ptr()
    };

    let ret = unsafe {
        crate::internal::peer_send(
            peer_ids_buf.as_ptr(),
            peer_ids.len() as u32,
            data_ptr,
            data.len() as u32,
        )
    };

    if ret < 0 {
        return Err(PeerError::from_send_code(ret));
    }

    Ok(())
}

/// A synchronous iterator over peer messages. The caller is responsible for
/// managing polling frequency (e.g., with sleep)
pub struct MessageReader {
    from_peer_buf: [u8; PeerId::BYTE_LEN],
    data_buf: Vec<u8>,
}

impl MessageReader {
    pub fn new() -> Self {
        MessageReader {
            from_peer_buf: [0u8; PeerId::BYTE_LEN],
            data_buf: vec![0u8; 64 * 1024],
        }
    }
}

impl Default for MessageReader {
    fn default() -> Self {
        Self::new()
    }
}

impl Iterator for MessageReader {
    type Item = Message;

    fn next(&mut self) -> Option<Self::Item> {
        let ret = unsafe {
            crate::internal::peer_recv(
                self.from_peer_buf.as_mut_ptr(),
                self.data_buf.as_mut_ptr(),
                self.data_buf.len() as u32,
            )
        };

        if ret > 0 {
            let from_peer = PeerId::try_from(self.from_peer_buf)
                .expect("host returned invalid peer ID bytes");
            let data = self.data_buf[..ret as usize].to_vec();
            Some(Message {
                from: from_peer,
                data,
            })
        } else {
            // ret == 0: no data available immediately
            // ret < 0: error (treat as no data for iterator pattern)
            None
        }
    }
}
